fxsql/README.md

```diff
@@ -0,0 +1,269 @@
+- `statement:exec-context`
+- `statement:query`
+- `statement:query-context`
+- `transaction:commit`
+- `transaction:rollback`
+
+### Migrations
+
+This module provides the possibility to run your DB schemas migrations, using [Goose](https://github.com/pressly/goose) under the hood.
+
+First, configure the path for your migration files:
+
+```yaml
+# ./configs/config.yaml
+modules:
+  sql:
+    migrations: db/migrations
+```
+
+You can then create a migration file in this path.
+
+For example, `db/migrations/00001_create_foo_table.sql`:
+
+```sql
+-- +goose Up
+CREATE TABLE foo (
+    id   INTEGER NOT NULL PRIMARY KEY,
+    bar VARCHAR(255)
+);
+
+-- +goose Down
+DROP TABLE IF EXISTS foo;
+```
+
+You can then run the migrations with `RunFxSQLMigration`, by specifying a migration command:
+
+```go
+package main
+
+import (
+	"database/sql"
+
+	"github.com/ankorstore/yokai/fxconfig"
+	"github.com/ankorstore/yokai/fxlog"
+	"github.com/ankorstore/yokai/fxsql"
+	"github.com/ankorstore/yokai/fxtrace"
+	"go.uber.org/fx"
+)
+
+func main() {
+	fx.New(
+		fxconfig.FxConfigModule,       // load the module dependencies
+		fxlog.FxLogModule,
+		fxtrace.FxTraceModule,
+		fxsql.FxSQLModule,             // load the module
+		fxsql.RunFxSQLMigration("up"), // run migration command "up"
+	).Run()
+}
+```
+
+Available migration commands:
+
+- `up`: migrate the DB to the most recent version available
+- `up-by-one`: migrate the DB up by 1
+- `up-to VERSION`: migrate the DB to a specific VERSION
+- `down`: roll back the version by 1
+- `down-to VERSION`: roll back to a specific VERSION
+- `redo`: re-run the latest migration
+- `reset`: roll back all migrations
+- `status`: dump the migration status for the current DB
+- `version`: print the current version of the database
+- `create NAME [sql|go]`: creates new migration file with the current timestamp
+- `fix`: apply sequential ordering to migrations
+- `validate`: check migration files without running them
+
+If you want to automatically shut down your Fx application after the migrations, you can use `RunFxSQLMigrationAndShutdown`.
+
+### Hooks
+
+This module provides the possibility to register several [Hook](https://github.com/ankorstore/yokai/blob/main/sql/hook.go) implementations to `extend the logic around the SQL operations`.
+
+This is done via the `AsSQLHook()` function:
+
+```go
+package main
+
+import (
+	"context"
+
+	"github.com/ankorstore/yokai/fxconfig"
+	"github.com/ankorstore/yokai/fxlog"
+	"github.com/ankorstore/yokai/fxsql"
+	"github.com/ankorstore/yokai/fxtrace"
+	"github.com/ankorstore/yokai/sql"
+	"go.uber.org/fx"
+)
+
+// example SQL hook
+type ExampleHook struct{}
+
+func NewExampleHook() *ExampleHook {
+	return &ExampleHook{}
+}
+
+func (h *ExampleHook) Before(ctx context.Context, event *sql.HookEvent) context.Context {
+	// before SQL operation logic
+
+	return ctx
+}
+
+func (h *ExampleHook) After(ctx context.Context, event *sql.HookEvent) {
+	// after SQL operation logic
```

Comment:
Why not returning the context here as well, as in before? (genuine question)

Reply:
I think it's the most simple example. It's supposed to return a new context if you wanna add something to the context. :)

Reply:
Yes, it works this way inside an operation:

- ctx = hook.Before(ctx) // may extend the CTX
- operation(ctx)
- hook.After(ctx)

Nothing is ever done after hook.After(), there is no point to return the ctx.